FEAT add DeLoRA

mwbini opened the pull request:
I implemented a first version, based on #2760

BenjaminBossan commented on src/peft/tuners/delora/config.py:
```diff
@@ -0,0 +1,147 @@
+
+        # check for layers_to_transform and layers_pattern
+        if self.layers_pattern and not self.layers_to_transform:
+            raise ValueError("When `layers_pattern` is specified, `layers_to_transform` must also be specified. ")
```
Would it make sense to check if `init_weights` and `use_residual_init` are compatible?

BenjaminBossan commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,630 @@
+        "delora_w_norm",
+        "delora_initial_w_norm",
+        # Misc flag
+        "use_residual_init",
```
I think there is no need for `"use_residual_init"` here.

BenjaminBossan commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,630 @@
+        self.delora_B = nn.ParameterDict({})
+        self.delora_alpha = nn.ParameterDict({})
+        # Store flag from kwargs (defaults to True if not provided)
+        self.use_residual_init = bool(kwargs.get("use_residual_init", True))
```
Let's not do `kwargs.get(...)`, just put `use_residual_init: bool = True` into the `__init__` arguments.

BenjaminBossan commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,630 @@
+        delta = delta * w_norm_.unsqueeze(0)
+        return delta.to(A.dtype, copy=False)
+
+    def get_delta_weight(self, adapter: str) -> torch.Tensor:
```
There is a lot of similarity between this method and `_compute_delta_from_params`, I wonder if, we a little refactoring of `_compute_delta_from_params`, it could be reused here. WDYT?

mwbini replied:
Thanks for the suggestion, I refactored the code to reuse the same `_compute_delta` function

BenjaminBossan commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,630 @@
+        diag = torch.diag_embed((alpha / self.r[adapter]) / (An * Bn))  # (r,r)
+        delta = B @ diag @ A  # (out,in)
+
+        if (not self.use_residual_init) and (adapter in self.delora_initial_A and adapter in self.delora_initial_B):
```
Could you please add a comment about why this conditional is needed?

BenjaminBossan commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,630 @@
+        if adapter in self.delora_w_norm:
+            w_norm = self.delora_w_norm[adapter].to(dtype=comp_dtype, device=device)
+        else:
+            # fallback: compute current base weight norm on-the-fly
```
Under what conditions could `self.delora_w_norm` not contain the adapter?

mwbini replied:
Indeed this was mistakenly still in the code. I now removed it

BenjaminBossan commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,630 @@
+        self.r[adapter_name] = r
+        if use_residual_init is not None:
+            # allow toggling via update for already-wrapped layers
+            self.use_residual_init = bool(use_residual_init)
```
I don't think we need casting to `bool`, the argument is documented as being of `bool` type and we generally should assume that users pass the correct type (otherwise, we'd have to carefully check each possible argument).

BenjaminBossan commented on src/peft/tuners/delora/config.py:
```diff
@@ -0,0 +1,147 @@
+            Bias type for DeLoRA. Can be 'none' or 'all'. If 'all', the corresponding biases will be updated during
+            training. Be aware that this means that, even when disabling the adapters, the model will not produce the
+            same output as the base model would have without adaptation.
+        init_weights (`bool`):
```
There are more options than just `True` or `False` here, right? Let's document all of them.

BenjaminBossan commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,630 @@
+                        alpha0 = self.delora_initial_alpha[adapter_name].to(dtype=w.dtype, device=w.device).squeeze(0)
+                        wnorm0 = self.delora_initial_w_norm[adapter_name].to(dtype=w.dtype, device=w.device)
+                        delta0 = self._compute_delta_from_params(A0, B0, alpha0, self.r[adapter_name], wnorm0)
+                        self.get_base_layer().weight.data.sub_(delta0)
```
If we add two DeLoRA adapters with `use_residual_init=True` to the model, would the 2nd one not subtract its delta again from the base weight? Would this not result in the base weight being incorrect for the first adapter?

BenjaminBossan commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,630 @@
+        # This should not be done for a model that has just been loaded from a checkpoint, as the state is already correct.
+        if self.use_residual_init and len(self._initials_applied) > 0:
+            w = self.get_base_layer().weight
+            for adapter in list(self._initials_applied):
```
No need to cast to list.

BenjaminBossan commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,630 @@
+                    wnorm0 = self.delora_initial_w_norm[adapter].to(dtype=w.dtype, device=w.device)
+                    delta0 = self._compute_delta_from_params(A0, B0, alpha0, self.r[adapter], wnorm0)
+                    self.get_base_layer().weight.data = self.get_base_layer().weight.data + delta0
+                self._initials_applied.discard(adapter)
```
Since we checked earlier that `adapter` is in `_initials_applied`, there is no need to use `discard`, you can use `remove`.

BenjaminBossan commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,630 @@
+            base_out = self.base_layer(x, *args, **kwargs)
+            add_out = torch.zeros_like(base_out)
+
+            for a in self.active_adapters:
```
Let's not use `a` but `adapter` or `adapter_name`.

BenjaminBossan reviewed (changes requested):
Thanks a lot for this PR to add DeLoRA to PEFT. The PR is already quite advanced and in a nice state, good job with that.

I did a review of the PR and added a few comments, please check those. But while working on this, I also noticed some more general areas for discussion, which should be addressed first:

## Naming

In general, I would suggest to rename variables with `DeLoRA` to `Delora` or `DeLora`. Not only is that easier to type, it's also consistent with the other names, e.g. `LoraConfig`.

## Initialization

I think a lot of complexity of the implementation comes from the initialization with `use_residual_init=True`. This is the reason why we need the `delora_initial_*` parameters and update methods like `_load_from_state_dict`, right? I wonder if we can think of a way to significantly reduce this complexity.

First of all, I still need to get a good understanding why this is required in the first place. Do I get it right that normally, initializing DeLoRA cannot be done in a way that makes it an identity transform? So there is no choice of `delora_A` and `delora_B` that would make their contribution zero? For me, this would be the ideal solution.

Second, how important is it to have the `use_residual_init=True` option? Is it better for training or could we just drop it entirely? I know that this would require some adjustments to the unit tests, as some tests assume that the adapter is an identity transform without training. But these adjustments can be done and it is not a requirement for PEFT methods to have an initialization option that comes down to identity.

If this option is not strictly needed for performance reasons but mainly to satisfy tests, I would rather remove it, eliminate a lot of the complexity associated with it, and then update the tests.

## Device and dtype handling

There is a lot of code that deals with moving the DeLoRA parameters to the correct device and dtype, such as:

```python
A = self.delora_A[a].to(dtype=comp_dtype, device=device)
B = self.delora_B[a].to(dtype=comp_dtype, device=device)
alpha = self.delora_alpha[a].to(dtype=comp_dtype, device=device)
```

Would it not make sense to initialize them with the correct device and dtype right from the start (as is normally done for other PEFT methods) instead of casting and moving them each time, which can be costly? Or is this some sort of memory optimization?

## Other

Those are the big points I found while reviewing and they should be the focus for the next iteration. For the final PR, I would also like to see an example (it could be a copy of an existing example with modifications to use DeLoRA) and setting for the [PEFT method comparison suite](https://github.com/huggingface/peft/tree/main/method_comparison/MetaMathQA).

Finally, please always call `make style` before pushing your changes.

mwbini commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,630 @@
+                    delta_weight = self.get_delta_weight(active_adapter)
+                    self.get_base_layer().weight.data.sub_(delta_weight)
+        # If we had subtracted initial delta(s) at setup, restore them now to get back the original base weights
+        # This should not be done for a model that has just been loaded from a checkpoint, as the state is already correct.
```
I see. I first added this in case one wanted to unmerge and revert back to base weights. 

Since this might be a secondary usage, I now added an additional argument `unapply_initials` to unmerge (which by default is False) that can further revert to base weights. I am not sure if this is necessary, and I am okay with removing it if it's not

mwbini commented:
Thank you for the feedbacks. Now I should have addressed most of the comments, please let me know if something is missing.

Regarding the initialization I am not sure what could be the best approach. Maybe B=0 could still be used by setting a large enough min value in torch.clamp... As a next step, I will try to setup some experiments and check this zero initialization, the overhead of using use_residual_init=False, etc.

Regarding using multiple adapters, it would be indeed not possible with use_residual_init=True. For now, I added a check in _check_new_adapter_config that should check for this

BenjaminBossan commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,610 @@
+            if adapter_name in self.delora_initial_lambda:
+                del self.delora_initial_lambda[adapter_name]
+
+    def _move_adapter_to_device_of_base_layer(self, adapter_name: str, device: Optional[torch.device] = None) -> None:
```
It seems that we cannot use the `_move_adapter_to_device_of_base_layer` method from the parent class (`BaseTunerLayer`). Could you please add a short comment on what is different here?

BenjaminBossan commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,610 @@
+        """
+        if device is None:
+            base_layer = self.get_base_layer()
+            if isinstance(base_layer, nn.MultiheadAttention):
+                base_layer = base_layer.out_proj
```
MHA is not supported by DeLoRA so this can be removed.

BenjaminBossan commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,610 @@
+        # Create a filtered state dict with only keys that belong to this module
+        module_state_dict = {}
+        for key, value in state_dict.items():
+            if key.startswith(prefix):
+                # Remove the prefix to get the relative key
+                relative_key = key[len(prefix) :]
```
`relative_key = key.removeprefix(prefix)` should do the job.

BenjaminBossan commented on src/peft/tuners/delora/layer.py:
```diff
@@ -0,0 +1,610 @@
+                module_state_dict[relative_key] = value
+
+        # Manually load ParameterDict and BufferDict parameters
+        for param_name in ["delora_A", "delora_B", "delora_lambda"]:
```
We can use `self.adapter_layer_names` here, right?

BenjaminBossan commented on src/peft/tuners/delora/model.py:
```diff
@@ -0,0 +1,300 @@
+        Raise a ValueError if there is something wrong with the config or if it conflicts with existing adapters.
+
+        """
+        # TODO: there should be a check if any of the existing adapters actually has bias != "none", or else the check
+        # does not fully correspond to the error message.
+        if (len(self.peft_config) > 1) and (config.bias != "none"):
+            raise ValueError(
+                f"{self.__class__.__name__} supports only 1 adapter with bias. When using multiple adapters, "
+                "set bias to 'none' for all adapters."
+            )
```
Let's replace this with:
```suggestion
        super()._check_new_adapter_config(config)
```

BenjaminBossan commented on src/peft/tuners/delora/model.py:
```diff
@@ -0,0 +1,300 @@
+                "set use_residual_init to False for all adapters."
+            )
+
+    @staticmethod
```
We recently had a PR (#2771) that allows us to greatly reduce the number of methods required to implement here. You can simply remove the following methods:

- `_check_target_module_exists`
- `_mark_only_adapters_as_trainable`
- `__getattr__`
- `get_peft_config_as_dict`
- `_set_adapter_layers`
- `enable_adapter_layers`
- `disable_adapter_layers`
- `set_adapter`
- `_prepare_adapter_config`
- `_unload_and_optionally_merge`
- `merge_and_unload`
- `unload`

On the flip side, you need to add:

```python
    tuner_layer_cls = DeloraLayer
    target_module_mapping = TRANSFORMERS_MODELS_TO_DELORA_TARGET_MODULES_MAPPING
```

as class attributes, i.e. directly below `prefix: str = "delora_"`.

For this to work, you also need to merge with/rebase on the latest main branch.

BenjaminBossan commented on tests/test_initialization.py:
```diff
@@ -1846,6 +1847,98 @@ def test_road_with_conv2d_layer(self):
+        assert tuple(B.shape) == (out_f, r)
+        assert tuple(lambda_.shape) == (1,)
+
+    def test_disable_enable_no_change(self, data):
+        base = self.get_model()
+        cfg = DeloraConfig(target_modules=["lin0"], r=8, lambda_=8)
+        model = get_peft_model(base, cfg)
+
+        y0 = model(data)
+        model.disable_adapter_layers()
+        y1 = model(data)
+        model.enable_adapter_layers()
+        y2 = model(data)
+
+        assert torch.allclose(y0, y1, atol=1e-6, rtol=1e-6)
+        assert torch.allclose(y0, y2, atol=1e-6, rtol=1e-6)
+
+    def test_merge_and_unload_same_outputs(self, data):
+        base = self.get_model()
+        cfg = DeloraConfig(target_modules=["lin0"], r=8, lambda_=8)
+        model = get_peft_model(base, cfg)
+        y_before = model(data)
+
+        merged = model.merge_and_unload()
+        merged.eval()
+        y_after = merged(data)
+        assert torch.allclose(y_before, y_after, atol=1e-6, rtol=1e-6)
```
I think this should already be well covered by the existing tests and can be removed here.

BenjaminBossan commented on tests/test_initialization.py:
```diff
@@ -1846,6 +1847,98 @@ def test_road_with_conv2d_layer(self):
+        model = get_peft_model(base, cfg)
+        y_peft = model(data)
+        assert torch.allclose(y_base, y_peft, atol=1e-6, rtol=1e-6)
+
```
Let's add a test adding a second adapter (via `model.add_adapter`) that has `use_residual_init=True` raises the "supports only 1 adapter with residual init" error.

BenjaminBossan reviewed (changes requested):
Thanks for the updates. I found a couple of new issues, but nothing big. Also, please merge with/rebase on the latest PEFT main branch and resolve any merge conflicts that may occur.

> Regarding the initialization I am not sure what could be the best approach. Maybe B=0 could still be used by setting a large enough min value in torch.clamp... As a next step, I will try to setup some experiments and check this zero initialization, the overhead of using use_residual_init=False, etc.

Great, LMK if you find something.

BenjaminBossan commented on examples/delora_finetuning/README.md:
```diff
@@ -0,0 +1,102 @@
+from transformers import AutoModelForCausalLM
+
+model = AutoModelForCausalLM.from_pretrained(
+    "meta-llama/Meta-Llama-3-8B", torch_dtype=torch.bfloat16, device_map="auto"
```
Let's use `dtype` instead of the deprecated `torch_dtype`.

BenjaminBossan reviewed (approved):
Thanks a lot for the last few changes and for contributing this method to PEFT. The PR is now good to go.
